### ISSUE TITLE
With llvm 14 and llvm 15 co-installed, meson doesn't pick llvm 15 using native-file

### ISSUE THREAD

**Issue by shmerl** [mesonbuild/meson#10483]
When trying to build Mesa using an override for default llvm, it doesn't pick llvm 15 when llvm 14 is present in the system.

OS: Debian testing
Meson: 0.62.2

The way I try to override (which did work in the past) is like this:

```
   read -r -d '' native_config <<EOF                                                                                            
[binaries]
llvm-config = '/usr/bin/llvm-config-15'
EOF
```

This sets native_config variable to that. Then meson is called like this:

```
   LC_ALL=C.UTF-8 meson "$src_dir" \
...
--native-file=<(echo "$native_config") \
...
```

So in the past it was picking llvm fine, but now it falls back to llvm 14 ignoring the override. Did anything change in meson or llvm that requires something new for that?

**Comment by shmerl** [comment]
May be it's not the right way to use it? Instructions are from the Mesa documentation here:

https://docs.mesa3d.org/meson.html#llvm

**Comment by shmerl** [comment]
Opened Mesa bug as well in case it's a configuration issue: https://gitlab.freedesktop.org/mesa/mesa/-/issues/6738

**Comment by 1ace** [comment]
Looks like the list of llvm versions hasn't been kept up to date, it's two major versions behind:  
https://github.com/mesonbuild/meson/blob/master/mesonbuild/environment.py#L152

I wonder if it would be worth it to just get all the `{tool}-*`, sort them (as floats, to support 3.x), and pick the highest number? I don't see any other way to avoid meson being broken every time there's a new llvm release :confused:

**Comment by dcbaker** [comment]
Meson should ignore that if you have a native file. What's in the `$builddir/meson-logs/meson-log.txt`? There will be some extra debug spew in there about both the native file processing and the dependency processing for llvm

**Comment by 1ace** [comment]
Yes indeed, my bad; this is actually not used in this case :see_no_evil::sweat_smile:

**Comment by dcbaker** [comment]
Ah, okay. I see what's happening. When you set `shared_llvm` to disabled we try the `cmake` finder before the `config-tool` finder, and cmake finds llvm-14. You can either edit the meson.build and set `dependency('llvm', ..., method : 'config-tool')`, or you can set `meson configure -Dcmake_prefix_path=<prefix path for llvm-15>` so that the cmake finder will get the right cmake.

**Comment by shmerl** [comment]
I see.

@1ace: Would it make sense to propose to update Mesa's meson.build like above?

**Comment by shmerl** [comment]
@dcbaker: Will Meson fall back to cmake method if let's say config-tool doesn't produce a result?

**Comment by eli-schwartz** [comment]
If you specifically say to use config-tool, then no.

If you just don't want to use cmake dependencies at all, you can force the point by specifying cmake in your native-file, but setting its location as /bin/false -- it will then be detected to not work, so Meson won't try using it as a dependency lookup method no matter what. :)

**Comment by shmerl** [comment]
Well, I'd need to disable it for llvm only, in case it can be used for other purposes.

**Comment by shmerl** [comment]
Is there some variable that would indicate that native file was provided for llvm? Adding such conditional setting to config-tool in meson.build can make it minimally disruptive.

**Comment by shmerl** [comment]
@1ace: Do you have some suggestion how to update meson.build for that? May be adding some explicit config variable can help. I can open an MR.

**Comment by shmerl** [comment]
Somehow this just started working correctly for me without changing anything upstream.

**Comment by shmerl** [comment]
Looks like this resurfaced with llvm version bump to 16.

**Comment by shmerl** [comment]
> Is there some variable that would indicate that native file was provided for llvm?
> Adding such conditional setting to config-tool in meson.build can make it minimally disruptive.

So I guess there is no way to do the above programmatically?
